docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+
+The HTTP Gateway protocol has provisions to transfer further chunks of the body data from the canister to the HTTP Gateway, to overcome the message limit of the Internet Computer. This streaming protocol is independent of any possible streaming of data between the HTTP Gateway and the HTTP client. The gateway may assemble the response as a whole before passing it on, or pass the chunks on directly, on the TCP or HTTP level, as it sees fit. When the Gateway is certifying the response, it must not pass on uncertified chunks.
+
+If the `streaming_strategy` field of the `HttpResponse` is set, the HTTP Gateway then uses further query calls to obtain further chunks to append to the body:
+
+If the function reference in the callback field of the `streaming_strategy` is not a method of the given canister, the Gateway fails the request.
+
+Else, it makes a query call to the given method, passing the token value given in the `streaming_strategy` as the argument.
+
+That query method returns a `StreamingCallbackHttpResponse`. The body therein is appended to the body of the HTTP response. This is repeated as long as the method returns some token in the token field until that field is null.
+
+The type of the token value is chosen by the canister; the HTTP Gateway obtains the Candid type of the encoded message from the canister and uses it when passing the token back to the canister. This generic use of Candid is not covered by the Candid specification, and may not be possible in some cases (e.g. when using "future types"). Canister authors may have to use "simple" types.
+
+## Upgrade to Update Calls
+
+If the canister sets `upgrade = opt true` in the `HttpResponse` reply from `http_request`, then the Gateway ignores all other fields of the response. The Gateway performs an [update](https://internetcomputer.org/docs/current/references/ic-interface-spec#http-call) call to `http_request_update`, passing the same `HttpRequest` record as the argument, and uses the resulting response from `http_request_update` instead.
+
+The value of the `upgrade` field returned from `http_request_update` is ignored.
+
+## Legacy Response Verification
+
+Version 1 response verification only supports verifying a request path and response body pair with only one response per request path. This is quite restrictive in the number of scenarios it can support. For example, redirection or client-side caching is not safe since the status code and headers required to verify responses of that nature are not included in the certification. Upon a query call to a canister’s `http_request` method, a single malicious node or boundary node can modify these parts of the HTTP response, leading to the following issues:
+
+- For any dApp, a single malicious node on the subnet can maliciously redirect the user anywhere they like.
+- dApps cannot load the service worker when embedded within iFrames.
+- The use of redirects and cookies is unsafe as they can be manipulated by malicious nodes.
```

Comment:
Isn't this the same as the first point. I would probably only keep this one.

Reply:
Agreed, thanks!